### PR TITLE
Fix find_width_of_character_at_span bounds check

### LINKED ISSUE #48508: internal compiler error: assertion failed: bpos.to_usize() >= mbc.pos.to_usize() + mbc.bytes
I tried this code:

lib.rs:
```rust
#![feature(iterator_step_by)]
#![feature(ord_max_min)]
#![feature(test)]
#![feature(unboxed_closures)]
#![feature(try_from)]
#![feature(slice_patterns)]
#![feature(non_ascii_idents)]

pub mod precompute;
```

precompute.rs:
```rust
#[test]
fn it_works() {
    fn inner() -> f64 {
        let µ = 1.0;
        µ
    }
    let _µ = inner();
}
```

I expected to see this happen: compile properly and tests pass.

Instead, this happened: internal compiler error.

Possibly related to #46773, but I wasn't sure since the backtraces seemed rather different.

## Meta

`rustc --version --verbose`:
```
rustc 1.26.0-nightly (063deba92 2018-02-23)
binary: rustc
commit-hash: 063deba92e44809125a433ca6e6c1ad0993313bf
commit-date: 2018-02-23
host: x86_64-unknown-linux-gnu
release: 1.26.0-nightly
LLVM version: 6.0
```
Backtrace:
```
   Compiling unicode v0.1.0 (file:///home/jonathan/tmp/unicode)
thread 'rustc' panicked at 'assertion failed: bpos.to_usize() >= mbc.pos.to_usize() + mbc.bytes', libsyntax/codemap.rs:764:17
stack backtrace:
   0:     0x7f065582e063 - std::sys::unix::backtrace::tracing::imp::unwind_backtrace::hd00aa766a56fe293
                               at libstd/sys/unix/backtrace/tracing/gcc_s.rs:49
   1:     0x7f0655825554 - std::sys_common::backtrace::_print::h54da06fb16cb9ad6
                               at libstd/sys_common/backtrace.rs:71
   2:     0x7f065582ac9d - std::panicking::default_hook::{{closure}}::h67dd2c5bdfbcdba1
                               at libstd/sys_common/backtrace.rs:59
                               at libstd/panicking.rs:206
   3:     0x7f065582a999 - std::panicking::default_hook::h14daf183f7987b45
                               at libstd/panicking.rs:222
   4:     0x7f065582b0ff - std::panicking::rust_panic_with_hook::haf158391394f8248
                               at libstd/panicking.rs:400
   5:     0x7f06516247c5 - std::panicking::begin_panic::hc4b15280f4aac00e
   6:     0x7f06516ada1b - syntax::codemap::CodeMap::bytepos_to_file_charpos::h1904a2e0951f1ef0
   7:     0x7f06516aa4e4 - syntax::codemap::CodeMap::lookup_char_pos::hf465c34067aaa7fe
   8:     0x7f064aaf53f3 - rustc_trans::mir::FunctionCx::set_debug_loc::h1fab6c6c56da0887
   9:     0x7f064aafb4db - rustc_trans::mir::trans_mir::hd5460c3409d3184b
  10:     0x7f064aabd553 - rustc_trans::base::trans_instance::h7a0842c683f85f40
  11:     0x7f064aac8f4d - rustc_trans::base::compile_codegen_unit::h89925300458c236e
  12:     0x7f065241df64 - rustc::dep_graph::graph::DepGraph::with_task_impl::h8d57a8788dba82ff
  13:     0x7f06527f5b50 - rustc::ty::maps::<impl rustc::ty::maps::queries::compile_codegen_unit<'tcx>>::force::h9473f2f719c9e900
  14:     0x7f06527f6f2f - rustc::ty::maps::<impl rustc::ty::maps::queries::compile_codegen_unit<'tcx>>::try_get::hb1acd3d0f00d8e17
  15:     0x7f0652816a4c - rustc::ty::maps::TyCtxtAt::compile_codegen_unit::h0f13d4057b5fa559
  16:     0x7f065280f5b9 - rustc::ty::maps::<impl rustc::ty::context::TyCtxt<'a, 'tcx, 'lcx>>::compile_codegen_unit::h57dae7b07151dfca
  17:     0x7f064aabf637 - rustc_trans::base::trans_crate::h510f789f4dd2fca5
  18:     0x7f064ab1b007 - <rustc_trans::LlvmTransCrate as rustc_trans_utils::trans_crate::TransCrate>::trans_crate::hc34c86a7991a2357
  19:     0x7f0655be16c2 - rustc_driver::driver::phase_4_translate_to_llvm::heb37816963659b45
  20:     0x7f0655bda6ce - rustc_driver::driver::compile_input::{{closure}}::hb05e220533907e90
  21:     0x7f0655b7176e - rustc::ty::context::TyCtxt::create_and_enter::h74bf9fb47bfc54a3
  22:     0x7f0655bd867b - rustc_driver::driver::compile_input::hac1d19a7e4d66b92
  23:     0x7f0655bf575a - rustc_driver::run_compiler::h5fd8f6d44462a7b6
  24:     0x7f0655b20a8a - std::sys_common::backtrace::__rust_begin_short_backtrace::h093a629e65f72505
  25:     0x7f0655843c4e - __rust_maybe_catch_panic
                               at libpanic_unwind/lib.rs:102
  26:     0x7f0655b59ca6 - <F as alloc::boxed::FnBox<A>>::call_box::h41c53d796e08f220
  27:     0x7f065583b00b - std::sys::unix::thread::Thread::new::thread_start::h72150edde1f62ab4
                               at /checkout/src/liballoc/boxed.rs:798
                               at libstd/sys_common/thread.rs:24
                               at libstd/sys/unix/thread.rs:90
  28:     0x7f065070f7fb - start_thread
  29:     0x7f06554f3b5e - clone
  30:                0x0 - <unknown>
```

### PR DESCRIPTION
**Pull request by etaoins**
Commit 0bd96671f0 added bounds checking of our current target byte position to prevent infinite loops. Unfortunately it was comparing the file-relative `target` versus the global `file_start_pos` and `file_end_pos`.

The result is failing to detect multibyte characters unless their file-relative offset fit within their global offset. This causes other parts of the compiler to generate spans pointing to the middle of a
multibyte character which will ultimately panic in `bytepos_to_file_charpos`.

Fix by comparing the `target` to the total file size when moving forward and doing checked subtraction when moving backwards. This should preserve the intent of the bounds check while removing the offset confusion.

cc @davidtwco 

Fixes #48508

### CODE REVIEW

**Comment by estebank** [comment]
Could you add a test to catch any regressions?

**Comment by etaoins** [comment]
@estebank Added a new commit with a test

**Comment by Manishearth** [comment]
Fails pretty:

```
[01:46:36] failures:
[01:46:36] 
[01:46:36] ---- [pretty] run-pass/issue-48508.rs stdout ----
[01:46:36] 	
[01:46:36] error: pretty-printed source does not typecheck
[01:46:36] status: exit code: 101
[01:46:36] command: "/checkout/obj/build/x86_64-unknown-linux-gnu/stage2/bin/rustc" "-" "-Zno-trans" "--out-dir" "/checkout/obj/build/x86_64-unknown-linux-gnu/test/run-pass/issue-48508.pretty-out" "--target=x86_64-unknown-linux-gnu" "-L" "/checkout/obj/build/x86_64-unknown-linux-gnu/test/run-pass" "-L" "/checkout/obj/build/x86_64-unknown-linux-gnu/test/run-pass/issue-48508.stage2-x86_64-unknown-linux-gnu.pretty.aux" "-Crpath" "-O" "-Zmiri" "-Zunstable-options" "-Lnative=/checkout/obj/build/x86_64-unknown-linux-gnu/native/rust-test-helpers" "-g"
[01:46:36] stdout:
[01:46:36] ------------------------------------------
[01:46:36] 
[01:46:36] ------------------------------------------
[01:46:36] stderr:
[01:46:36] ------------------------------------------
[01:46:36] error[E0425]: cannot find function `other` in module `other_file`
[01:46:36]   --> <anon>:25:25
[01:46:36]    |
[01:46:36] 25 | fn main() { other_file::other(); }
[01:46:36]    |                         ^^^^^ not found in `other_file`
[01:46:36] 
[01:46:36] error: aborting due to previous error
[01:46:36] 
[01:46:36] If you want more information on this error, try using "rustc --explain E0425"
[01:46:36] 
[01:46:36] ------------------------------------------
[01:46:36] 
[01:46:36] thread '[pretty] run-pass/issue-48508.rs' panicked at 'explicit panic', tools/compiletest/src/runtest.rs:2893:9
[01:46:36] note: Run with `RUST_BACKTRACE=1` for a backtrace.
[01:46:36] 
[01:46:36] 
[01:46:36] failures:
[01:46:36]     [pretty] run-pass/issue-48508.rs
[01:46:36] 
[01:46:36] test result: [31mFAILED(B[m. 2879 passed; 1 failed; 49 ignored; 0 measured; 0 filtered out
[01:46:36] 
[01:46:36] thread 'main' panicked at 'Some tests failed', tools/compiletest/src/main.rs:476:22
```

**Comment by etaoins** [comment]
@Manishearth That looks like a bug in pretty, no? The test compiles and runs correctly so I'm not sure why pretty can't typecheck it.

Is it worth committing this without the test?

cc @estebank

File: src/test/run-pass/issue-48508-aux.rs [hunk 1]
```diff
@@ -0,0 +1,16 @@
+// Copyright 2018 The Rust Project Developers. See the COPYRIGHT
```
Comment (Manishearth) [inline src/test/run-pass/issue-48508-aux.rs]: Shouldn't this file be in the auxiliary folder?

**Comment by Manishearth** [comment]
Perhaps. The `path` flag may not work if the pretty file is compiled elsewhere. `ignore-pretty` is an acceptable fix

File: src/test/run-pass/issue-48508-aux.rs (hunk 1 above)
Comment (etaoins) [reply src/test/run-pass/issue-48508-aux.rs]: Everything in `auxillary` is treated as a crate. This needs to be a module to trigger #48508

Comment (Manishearth) [reply src/test/run-pass/issue-48508-aux.rs]: Ah, ok

**Pull request merged**